Add a wait for ready user guide as part of 2023Q2 Doc Fixit (#1134)
grpc/grpc.io: content/en/docs/guides/wait-for-ready.md

---
title: Wait-for-Ready
description: >-
  Explains how to configure RPCs to wait for the server to be ready before sending the request.
---

### Overview

This is a feature which can be used on a stub which will cause the RPCs to wait
(until an optional deadline is exceeded) for the server to become available
before sending the request.  This allows for robust batch workflows as transient server problems
(including startup) won't cause failures.

When an RPC is created while the channel is not in a READY state, without
waitForReady it will immediately return a failure; with waitForReady it will
simply block until the connection becomes ready.  The default is **without**
Wait-for-Ready.

For detailed semantics see [this][grpc doc].

### How to use Wait-for-Ready

You can specify for a stub whether or not it should use Wait-for-Ready which
will automatically be passed along, when an RPC is created, to the associated
stream.

{{% alert title="Note" color="info" %}}
 The RPC can still fail for other reasons besides the server not being
ready, so error handling is still necessary.
{{% /alert %}}

> The following shows the sequence of events the occur, when a client sends a

Reviewer: "the" -> "that"

Author: done

---
title: Wait For Ready
description: >-

>   Explains how and why to configure the wait for ready feature.

Reviewer 2: As is it seems a bit redundant. Maybe
```suggestion
  Explains how to configure rpcs to wait for the server to be ready before sending the request.
```

Author: done.

---

### Overview

> This is a feature which can be used on a stub which will cause the rpcs to wait

Reviewer 3: Style: "rpcs" -> "RPCs"

Author: done

> This is a feature which can be used on a stub which will cause the rpcs to wait

Reviewer 4: s/rpcs/RPCs/

> (until an optional deadline is exceeded) for the server to become available

Reviewer 5: Move the part about the deadline out of a parenthetical. Just make it another sentence, that "the deadline still applies."

Author: Done.

> (until an optional deadline is exceeded or) for the server to become available

Reviewer: Did you mean to leave the "or"?

Author: Fixed

before sending the request.

> This allows for robust batch workflows as transient server problems (including

Reviewer 5: Is there a reason start-up was called out?

Author: Reflects my earlier thinking that we were not waiting if in IDLE or CONNECTING state.  I've removed it.

> startup) won't cause job failures.

Reviewer 4: Should we say "job failures"  or "batch failures"?

Author: Removed "job" so it is now just "failures" which is more accurate; the call failed, not necessarily the whole process.

> allows simplifying the implementation by hiding transient failures.

Reviewer: Maybe better phrased as below?

"This allows for robust batch workflows as transient server problems won't cause job failures".

Author: done

> When an RPC is created while the channel is not in a READY state, without

Reviewer 5: If a channel is IDLE or CONNECTING, the RPC won't fail. May be best to say something more like "failed to connect to the server."

Author: done

> waitForReady it will immediately return a failure; with waitForReady it will

Author: It will depend on precisely what type of transient failure occurred.  This has been slightly rewritten.  What actually happens is that when an RPC comes along the channel tries to connect to the server and it is the failure of that which puts/keeps the subchannel and then the channel into/in TRANSIENT_FAILURE state.  The error that is returned has the details which are provided to the caller.

> simply block until the connection becomes ready.

Reviewer 5: s/block/be queued/

Author: done

> For detailed semantics see: [grpc/doc/wait-for-ready.md][grpc doc]

Reviewer 3: Maybe just "For detailed semantics, see this." and link the word "this"? It's not clear what significance the current link text has.

Author: done

### How to use Wait for Ready

> waitForReady is an attribute associated with a stub. When an RPC is created, the

Reviewer 5: This is a bit language-specific. We can just say, "You configure an RPC to be wait-for-ready when it is being started" or some such. ("You can tell gRPC an RPC will be wait-for-ready", etc)

Author: **Don't all languages have stubs?**  

I think the fact that the configuration is applied to the stub is an important point, but if not all languages use stubs then this is the wrong place for it.

I changed it to 
```
You can specify for a stub whether or not it should use Wait-for-Ready which
will automatically be passed along, when an RPC is created.
```

> waitForReady is an attribute associated with a stub. When an RPC is created, the

Reviewer 2: ```suggestion
`waitForReady` is an attribute associated with a stub. When an RPC is created, the
```

> waitForReady is an attribute associated with a stub. When an RPC is created, the

Reviewer 3: This is definitely going to be a debate, but I think the styling "wait-for-ready" might make more sense in English prose.

> wait for ready attribute is automatically provided to the associated stream.

Reviewer 3: Let's be consistent on the styling of the term everywhere.

> Warning: the RPC can still fail for other reasons besides the server not being

Reviewer 2: A simple note should suffice.
```suggestion
Note: the RPC can still fail for other reasons besides the server not being
```

Author: Done

> Warning:  the RPC can still fail for other reasons than the server not being

Reviewer: Extra space.

Author: Did you know that the convention to use 1 space instead of 2 after a colon happened in 1989?  I'm 34 years late updating my style guide :)

> Warning:  the RPC can still fail for other reasons than the server not being

Reviewer: "reasons than the server" -> "reasons besides the server" ?

Author: done

ready, so you must still implement error handline.

```mermaid
flowchart TD
SENT
F(Failure Status)
CH(Channel) -->|Create stub &\n set waitForReady| S(Stub)
S --> |create RPC| RPC
RPC -->|Initiates Communication| CH
CH --> ST{Channel State}
ST --> | READY| SENT(RPC Sent to server)

> ST -->|TRANSIENT_FAILURE / IDLE| BUR(Block on channel becoming READY or a parmanent failure)

Reviewer: "parmanent" -> "permanent"

  BUR --> |Channel becomes READY| SENT
  BUR --> |Permanent Failure / \nSHUTDOWN| F
  ST --> |Permanent Failure / \nSHUTDOWN| F
```

### Alternatives

> - Loop until RPC stops returning transient failures.

Reviewer 5: with exponential backoff

Author: done

> - Implement an onReady Handler and handle your own blocking _(for languages that

Reviewer 5: This isn't really an alternative.

Author: Good point. I changed it to say that using an onReady handler could be an optimization.   Added accepting the failures that might have been avoided in order to fail fast.

> - Implement an onReady Handler and handle your own blocking _(for languages that

Reviewer 2: ```suggestion
- Implement an `onReady` Handler and handle your own blocking _(for languages that
```

Author: done
